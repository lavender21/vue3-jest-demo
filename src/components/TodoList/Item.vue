<template>
  <div>
    <span class="todo-item" :class="dynimacClass" @click="setItemCompleted">
      {{ title }}
    </span>
    <span class="remove-btn" @click="removeItem">remove</span>
  </div>
</template>
<script lang="ts" setup>
import { computed } from "vue";

const props = defineProps<{ title: string; completed: boolean }>();

const dynimacClass = computed(() => (props.completed ? "completed" : ""));

const emit = defineEmits(["remove", "complete"]);

const removeItem = () => {
  emit("remove");
};
const setItemCompleted = () => {
  emit("complete");
};
</script>
<style lang="scss" scoped></style>
