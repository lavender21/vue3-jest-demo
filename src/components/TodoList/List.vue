<template>
  <div class="todo-list-wrapper">
    <PersonCenter userId="007" />
    <div class="input-item-wrapper">
      <input type="text" v-model="inputValue" class="item-input" />
      <button class="add-todo-btn" @click="addItem">Add</button>
    </div>
    <Item
      v-for="(item, index) in todoItems"
      :key="`item${index}`"
      :title="item.title"
      :completed="item.completed"
      @toggle="toggleItem(index)"
      @remove="removeItem(index)"
    />
  </div>
</template>
<script lang="ts" setup>
import { ref } from "vue";
import Item from "./Item.vue";
import PersonCenter from "../PersonCenter.vue";
const props = defineProps<{ todos: [{ title: string; completed: boolean }] }>();

const todoItems = ref(props.todos);
const inputValue = ref("");
const removeItem = (index: number) => {
  todoItems.value.splice(index, 1);
};
const addItem = () => {
  todoItems.value.push({
    title: inputValue.value,
    completed: false,
  });
  inputValue.value = "";
};

const toggleItem = (index: number) => {
  const currentStatus = todoItems.value[index].completed;
  todoItems.value[index].completed = !currentStatus;
};
</script>
<style lang="scss" scoped>
.todo-list-wrapper {
  width: 500px;
  margin: 20px auto;
}
.input-item-wrapper {
  margin: 8px;
  width: 100%;
  display: flex;
  justify-content: space-between;
}
.item-input {
  padding: 8px;
  flex: 1;
  border: 1px solid #ddd;
  border-radius: 0;
}
.add-todo-btn {
  padding: 8px;
  width: 60px;
  background: #3fb983;
  color: #fff;
  font-weight: 700;
  border: none;
}
</style>
