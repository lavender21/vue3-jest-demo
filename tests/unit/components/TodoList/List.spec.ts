import { mount, shallowMount } from "@vue/test-utils";
import List from "@/components/TodoList/List.vue";
import Item from "@/components/TodoList/Item.vue";

describe("List component, component UI test demo", function () {
  //  输入： List [{title: completed:}]
  const shallowMountTodoList = () => {
    const defualtProps = {
      todos: [
        { title: "准备session ppt", completed: true },
        { title: "准备session demo code", completed: false },
      ],
    };
    return shallowMount(List, {
      props: defualtProps,
    });
  };
  it("should render list item", function () {
    const wrapper = shallowMountTodoList();

    expect(wrapper.findAllComponents(Item).length).toEqual(2);
  });
  // 测试契约部分的输出
  it("should pass title and completed props to item", function () {
    const wrapper = shallowMountTodoList();

    expect(wrapper.findComponent(Item).props()).toEqual({
      title: "准备session ppt",
      completed: true,
    });
  });
  // 测试监听Vue自定义事件
  it("should remove current item from list when Item component emit remove event", async function () {
    const wrapper = shallowMountTodoList();

    await wrapper.findAllComponents(Item)[0].vm.$emit("remove");

    expect(wrapper.findAllComponents(Item)[0].props()).toEqual({
      title: "准备session demo code",
      completed: false,
    });
  });

  it("should add item when input title and click add button", async function () {
    const wrapper = shallowMountTodoList();

    await wrapper.find('input[type="text"]').setValue("session rehearsal");
    await wrapper.find("button.add-todo-btn").trigger("click");
    expect(wrapper.findAllComponents(Item)[2].props()).toEqual({
      title: "session rehearsal",
      completed: false,
    });
  });

  it("should toggle item when Item component emit toggleComplete event", async function () {
    const wrapper = shallowMountTodoList();
    const itemWrapper = wrapper.findAllComponents(Item)[1];
    await itemWrapper.vm.$emit("toggle");

    expect(itemWrapper.props()).toEqual({
      title: "准备session demo code",
      completed: true,
    });

    await itemWrapper.vm.$emit("toggle");
    expect(itemWrapper.props()).toEqual({
      title: "准备session demo code",
      completed: false,
    });
  });

  // mount 和 shallowMount对比 snapshot 要等组件完成最后再加
  xit("should render list item with mount", function () {
    const wrapper = shallowMountTodoList();

    expect(wrapper.html()).toMatchSnapshot();
  });
  xit("should render list item with shallowMount", function () {
    const wrapper = shallowMountTodoList();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
