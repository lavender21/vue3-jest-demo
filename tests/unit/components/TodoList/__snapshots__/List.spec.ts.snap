// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component, component UI test demo should render list item with mount 1`] = `
<div class="todo-list-wrapper">
  <div class="input-item-wrapper"><input type="text" class="item-input"><button class="add-todo-btn">Add</button></div>
  <div class="todo-item-wrapper"><span class="todo-item completed">准备session ppt</span><span class="remove-btn">remove</span></div>
  <div class="todo-item-wrapper"><span class="todo-item">准备session demo code</span><span class="remove-btn">remove</span></div>
</div>
`;

exports[`List component, component UI test demo should render list item with shallowMount 1`] = `
<div class="todo-list-wrapper">
  <div class="input-item-wrapper"><input type="text" class="item-input"><button class="add-todo-btn">Add</button></div>
  <item-stub title="准备session ppt" completed="true"></item-stub>
  <item-stub title="准备session demo code" completed="false"></item-stub>
</div>
`;
